File: tests/wasm/test_cksp.js
/**
 * WASM binding tests for CKSP indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/my_project.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('CKSP partial params', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    
    const result = wasm.cksp_js(high, low, close, 10, 1.0, 9);
    assert.strictEqual(result.length, close.length * 2);
});

test('CKSP accuracy', async () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    
    const expected = EXPECTED_OUTPUTS['cksp'];
    const params = expected['default_params'];
    
    
    const result = wasm.cksp_js(high, low, close, params.p, params.x, params.q);
    
    
    assert.strictEqual(result.length, close.length * 2);
    
    
    const len = close.length;
    const longResult = result.slice(0, len);
    const shortResult = result.slice(len);
    
    
    const expectedLongLast5 = [
        60306.66197802568,
        60306.66197802568,
        60306.66197802568,
        60203.29578022311,
        60201.57958198072,
    ];
    const longLast5 = longResult.slice(-5);
    expectedLongLast5.forEach((expectedVal, i) => {
        assertClose(longLast5[i], expectedVal, 1e-5, `CKSP long value mismatch at index ${i}`);
    });
    
    
    const expectedShortLast5 = [
        58757.826484736055,
        58701.74383626245,
        58656.36945263621,
        58611.03250737258,
        58611.03250737258,
    ];
    const shortLast5 = shortResult.slice(-5);
    expectedShortLast5.forEach((expectedVal, i) => {
        assertClose(shortLast5[i], expectedVal, 1e-5, `CKSP short value mismatch at index ${i}`);
    });
    
    
    
    
    
});

test('CKSP default candles', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    
    const result = wasm.cksp_js(high, low, close, 10, 1.0, 9);
    assert.strictEqual(result.length, close.length * 2);
});

test('CKSP zero period', () => {
    
    const high = new Float64Array([10, 11, 12]);
    const low = new Float64Array([9, 10, 10.5]);
    const close = new Float64Array([9.5, 10.5, 11]);
    
    assert.throws(() => {
        wasm.cksp_js(high, low, close, 0, 1.0, 9);
    }, /Invalid param/);
});

test('CKSP invalid x', () => {
    
    const high = new Float64Array([10, 11, 12]);
    const low = new Float64Array([9, 10, 10.5]);
    const close = new Float64Array([9.5, 10.5, 11]);
    
    
    assert.throws(() => {
        wasm.cksp_js(high, low, close, 10, NaN, 9);
    }, /Invalid param/);
    
    
    assert.throws(() => {
        wasm.cksp_js(high, low, close, 10, Infinity, 9);
    }, /Invalid param/);
});

test('CKSP invalid q', () => {
    
    const high = new Float64Array([10, 11, 12]);
    const low = new Float64Array([9, 10, 10.5]);
    const close = new Float64Array([9.5, 10.5, 11]);
    
    assert.throws(() => {
        wasm.cksp_js(high, low, close, 10, 1.0, 0);
    }, /Invalid param/);
});

test('CKSP period exceeds length', () => {
    
    const high = new Float64Array([10, 11, 12]);
    const low = new Float64Array([9, 10, 10.5]);
    const close = new Float64Array([9.5, 10.5, 11]);
    
    assert.throws(() => {
        wasm.cksp_js(high, low, close, 10, 1.0, 9);
    }, /Not enough data/);
});

test('CKSP very small dataset', () => {
    
    const high = new Float64Array([42]);
    const low = new Float64Array([41]);
    const close = new Float64Array([41.5]);
    
    assert.throws(() => {
        wasm.cksp_js(high, low, close, 10, 1.0, 9);
    }, /Not enough data/);
});

test('CKSP all NaN input', () => {
    
    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);
    
    assert.throws(() => {
        wasm.cksp_js(allNaN, allNaN, allNaN, 10, 1.0, 9);
    }, /Data is empty|No data/);
});

test('CKSP error handling', () => {
    
    
    assert.throws(() => {
        wasm.cksp_js(new Float64Array(), new Float64Array(), new Float64Array(), 10, 1.0, 9);
    }, /Data is empty/);
    
    
    assert.throws(() => {
        wasm.cksp_js(
            new Float64Array([1, 2, 3]),
            new Float64Array([1, 2]),  
            new Float64Array([1, 2, 3]),
            10, 1.0, 9
        );
    }, /Inconsistent/);
    
    
    const high = new Float64Array([10, 11, 12]);
    const low = new Float64Array([9, 10, 10.5]);
    const close = new Float64Array([9.5, 10.5, 11]);
    assert.throws(() => {
        wasm.cksp_js(high, low, close, 0, 1.0, 9);
    }, /Invalid param/);
    
});

test('CKSP nan handling', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    
    const expected = EXPECTED_OUTPUTS['cksp'];
    const params = expected['default_params'];
    
    const result = wasm.cksp_js(high, low, close, params.p, params.x, params.q);
    
    
    const len = close.length;
    const longResult = result.slice(0, len);
    const shortResult = result.slice(len);
    
    
    const warmupPeriod = params.p + params.q - 1;
    assert.strictEqual(warmupPeriod, 18, 'Expected warmup period of 18');
    for (let i = 0; i < warmupPeriod && i < longResult.length; i++) {
        assert(isNaN(longResult[i]), `Expected NaN in long warmup at index ${i}`);
        assert(isNaN(shortResult[i]), `Expected NaN in short warmup at index ${i}`);
    }
    
    
    if (longResult.length > 240) {
        for (let i = 240; i < longResult.length; i++) {
            assert(!isNaN(longResult[i]), `Found unexpected NaN in long at index ${i}`);
            assert(!isNaN(shortResult[i]), `Found unexpected NaN in short at index ${i}`);
        }
    }
});

test('CKSP fast API (cksp_into)', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    
    const len = close.length;
    const expected = EXPECTED_OUTPUTS['cksp'];
    const params = expected['default_params'];
    
    
    const longPtr = wasm.cksp_alloc(len);
    const shortPtr = wasm.cksp_alloc(len);
    
    try {
        
        wasm.cksp_into(
            high,
            low,
            close,
            longPtr,
            shortPtr,
            len,
            params.p,
            params.x,
            params.q
        );
        
        
        const longView = new Float64Array(wasm.__wasm.memory.buffer, longPtr, len);
        const shortView = new Float64Array(wasm.__wasm.memory.buffer, shortPtr, len);
        
        
        const longResult = Array.from(longView);
        const shortResult = Array.from(shortView);
        
        
        const safeResult = wasm.cksp_js(high, low, close, params.p, params.x, params.q);
        const safeLong = safeResult.slice(0, len);
        const safeShort = safeResult.slice(len);
        
        assertArrayClose(
            longResult,
            Array.from(safeLong),
            1e-10,
            'Fast API long values should match safe API'
        );
        
        assertArrayClose(
            shortResult,
            Array.from(safeShort),
            1e-10,
            'Fast API short values should match safe API'
        );
        
    } finally {
        
        wasm.cksp_free(longPtr, len);
        wasm.cksp_free(shortPtr, len);
    }
});

test('CKSP batch processing', async () => {
    
    const high = new Float64Array(testData.high.slice(0, 100)); 
    const low = new Float64Array(testData.low.slice(0, 100));
    const close = new Float64Array(testData.close.slice(0, 100));
    
    
    const config = {
        p_range: [10, 10, 0],
        x_range: [1.0, 1.0, 0.0],
        q_range: [9, 9, 0]
    };
    
    const result = wasm.cksp_batch(high, low, close, config);
    
    
    assert(result.long_values, 'Missing long_values in batch result');
    assert(result.short_values, 'Missing short_values in batch result');
    assert(result.combos, 'Missing combos in batch result');
    assert.strictEqual(result.rows, 1, 'Expected 1 row for single parameter set');
    assert.strictEqual(result.cols, 100, 'Expected 100 columns');
    
    
    const multiConfig = {
        p_range: [5, 15, 5],  
        x_range: [0.5, 1.5, 0.5],  
        q_range: [5, 10, 5]  
    };
    
    const multiResult = wasm.cksp_batch(high, low, close, multiConfig);
    
    
    assert.strictEqual(multiResult.rows, 18, 'Expected 18 parameter combinations');
    assert.strictEqual(multiResult.combos.length, 18, 'Expected 18 combo entries');
    assert.strictEqual(multiResult.long_values.length, 18 * 100, 'Expected correct long_values length');
    assert.strictEqual(multiResult.short_values.length, 18 * 100, 'Expected correct short_values length');
    
    
    assert.strictEqual(multiResult.combos[0].p, 5);
    assert.strictEqual(multiResult.combos[0].x, 0.5);
    assert.strictEqual(multiResult.combos[0].q, 5);
    
    
    assert.strictEqual(multiResult.combos[17].p, 15);
    assert.strictEqual(multiResult.combos[17].x, 1.5);
    assert.strictEqual(multiResult.combos[17].q, 10);
    
    
    for (let i = 0; i < multiResult.combos.length; i++) {
        const combo = multiResult.combos[i];
        const expectedWarmup = combo.p + combo.q - 1;
        const rowStart = i * 100;
        const rowData = multiResult.long_values.slice(rowStart, rowStart + 100);
        
        
        let nanCount = 0;
        for (const val of rowData) {
            if (isNaN(val)) {
                nanCount++;
            } else {
                break;
            }
        }
        
        
        assert(nanCount >= Math.min(expectedWarmup, 100), 
               `Row ${i}: Expected at least ${expectedWarmup} NaN values, got ${nanCount}`);
    }
});

test('CKSP batch metadata', () => {
    
    const high = new Float64Array(testData.high.slice(0, 50));
    const low = new Float64Array(testData.low.slice(0, 50));
    const close = new Float64Array(testData.close.slice(0, 50));
    
    
    const singleConfig = {
        p_range: [10, 10, 1],
        x_range: [1.0, 1.0, 0.1],
        q_range: [9, 9, 1]
    };
    
    const singleResult = wasm.cksp_batch(high, low, close, singleConfig);
    assert.strictEqual(singleResult.rows, 1, 'Single sweep should have 1 row');
    assert.strictEqual(singleResult.combos.length, 1, 'Single sweep should have 1 combo');
    
    
    const largeStepConfig = {
        p_range: [10, 12, 10],  
        x_range: [1.0, 1.0, 0],
        q_range: [9, 9, 0]
    };
    
    const largeStepResult = wasm.cksp_batch(high, low, close, largeStepConfig);
    assert.strictEqual(largeStepResult.rows, 1, 'Large step should only have 1 row');
    assert.strictEqual(largeStepResult.combos[0].p, 10, 'Should only have p=10');
});

test.after(() => {
    console.log('CKSP WASM tests completed');
});

File: tests/python/test_cksp.py
"""
Python binding tests for CKSP indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path


sys.path.insert(0, str(Path(__file__).parent.parent.parent / 'target/wheels'))

try:
    import my_project as ta_indicators
except ImportError:
    pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, assert_all_nan, assert_no_nan, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestCksp:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_cksp_partial_params(self, test_data):
        """Test CKSP with partial parameters (None values) - mirrors check_cksp_partial_params"""

        long_result, short_result = ta_indicators.cksp(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            p=10, x=1.0, q=9
        )
        assert len(long_result) == len(test_data['close'])
        assert len(short_result) == len(test_data['close'])

    def test_cksp_accuracy(self, test_data):
        """Test CKSP matches expected values from Rust tests - mirrors check_cksp_accuracy"""
        expected = EXPECTED_OUTPUTS['cksp']
        params = expected['default_params']


        long_result, short_result = ta_indicators.cksp(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            p=params['p'],
            x=params['x'],
            q=params['q']
        )


        assert len(long_result) == len(test_data['close'])
        assert len(short_result) == len(test_data['close'])


        expected_long_last_5 = [
            60306.66197802568,
            60306.66197802568,
            60306.66197802568,
            60203.29578022311,
            60201.57958198072,
        ]
        assert_close(
            long_result[-5:],
            expected_long_last_5,
            rtol=1e-8,
            atol=1e-5,
            msg="CKSP long values mismatch"
        )


        expected_short_last_5 = [
            58757.826484736055,
            58701.74383626245,
            58656.36945263621,
            58611.03250737258,
            58611.03250737258,
        ]
        assert_close(
            short_result[-5:],
            expected_short_last_5,
            rtol=1e-8,
            atol=1e-5,
            msg="CKSP short values mismatch"
        )






    def test_cksp_default_candles(self, test_data):
        """Test CKSP with default parameters - mirrors check_cksp_default_candles"""

        long_result, short_result = ta_indicators.cksp(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            10, 1.0, 9
        )
        assert len(long_result) == len(test_data['close'])
        assert len(short_result) == len(test_data['close'])

    def test_cksp_with_kernel(self, test_data):
        """Test CKSP with different kernel options - mirrors kernel tests"""
        params = EXPECTED_OUTPUTS['cksp']['default_params']


        long_scalar, short_scalar = ta_indicators.cksp(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            p=params['p'],
            x=params['x'],
            q=params['q'],
            kernel='scalar'
        )


        long_auto, short_auto = ta_indicators.cksp(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            p=params['p'],
            x=params['x'],
            q=params['q']
        )


        assert_close(long_scalar, long_auto, rtol=1e-10, msg="CKSP long kernel mismatch")
        assert_close(short_scalar, short_auto, rtol=1e-10, msg="CKSP short kernel mismatch")

    def test_cksp_zero_period(self):
        """Test CKSP fails with zero period - mirrors check_cksp_zero_period"""
        high = np.array([10.0, 11.0, 12.0])
        low = np.array([9.0, 10.0, 10.5])
        close = np.array([9.5, 10.5, 11.0])

        with pytest.raises(ValueError, match="Invalid param"):
            ta_indicators.cksp(high, low, close, p=0, x=1.0, q=9)

    def test_cksp_invalid_x(self):
        """Test CKSP fails with invalid x multiplier"""
        high = np.array([10.0, 11.0, 12.0])
        low = np.array([9.0, 10.0, 10.5])
        close = np.array([9.5, 10.5, 11.0])


        with pytest.raises(ValueError, match="Invalid param"):
            ta_indicators.cksp(high, low, close, p=10, x=float('nan'), q=9)





        with pytest.raises(ValueError, match="Invalid param"):
            ta_indicators.cksp(high, low, close, p=10, x=float('inf'), q=9)

    def test_cksp_invalid_q(self):
        """Test CKSP fails with invalid q parameter"""
        high = np.array([10.0, 11.0, 12.0])
        low = np.array([9.0, 10.0, 10.5])
        close = np.array([9.5, 10.5, 11.0])

        with pytest.raises(ValueError, match="Invalid param"):
            ta_indicators.cksp(high, low, close, p=10, x=1.0, q=0)

    def test_cksp_period_exceeds_length(self):
        """Test CKSP fails when period exceeds data length - mirrors check_cksp_period_exceeds_length"""
        high = np.array([10.0, 11.0, 12.0])
        low = np.array([9.0, 10.0, 10.5])
        close = np.array([9.5, 10.5, 11.0])

        with pytest.raises(ValueError, match="Not enough data"):
            ta_indicators.cksp(high, low, close, p=10, x=1.0, q=9)

    def test_cksp_errors(self):
        """Test error handling - additional edge cases"""

        with pytest.raises(ValueError, match="Data is empty"):
            ta_indicators.cksp(
                np.array([]),
                np.array([]),
                np.array([]),
                p=10, x=1.0, q=9
            )


        with pytest.raises(ValueError, match="Inconsistent"):
            ta_indicators.cksp(
                np.array([1.0, 2.0, 3.0]),
                np.array([1.0, 2.0]),
                np.array([1.0, 2.0, 3.0]),
                p=10, x=1.0, q=9
            )


        high = np.array([10.0, 11.0, 12.0])
        low = np.array([9.0, 10.0, 10.5])
        close = np.array([9.5, 10.5, 11.0])
        with pytest.raises(ValueError, match="Invalid param"):
            ta_indicators.cksp(high, low, close, p=0, x=1.0, q=9)


        with pytest.raises(ValueError, match="Not enough data"):
            ta_indicators.cksp(high, low, close, p=10, x=1.0, q=9)

    def test_cksp_very_small_dataset(self):
        """Test CKSP with single data point - mirrors check_cksp_very_small_dataset"""
        high = np.array([42.0])
        low = np.array([41.0])
        close = np.array([41.5])

        with pytest.raises(ValueError, match="Not enough data"):
            ta_indicators.cksp(high, low, close, p=10, x=1.0, q=9)

    def test_cksp_all_nan_input(self):
        """Test CKSP with all NaN values"""
        all_nan = np.full(100, np.nan)

        with pytest.raises(ValueError, match="Data is empty|No data"):
            ta_indicators.cksp(all_nan, all_nan, all_nan, p=10, x=1.0, q=9)

    def test_cksp_nan_handling(self, test_data):
        """Test CKSP handles NaN values correctly - mirrors check_cksp_nan_handling"""
        params = EXPECTED_OUTPUTS['cksp']['default_params']

        long_result, short_result = ta_indicators.cksp(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            p=params['p'],
            x=params['x'],
            q=params['q']
        )


        warmup_period = params['p'] + params['q'] - 1
        assert warmup_period == 18, f"Expected warmup period of 18, got {warmup_period}"
        assert_all_nan(long_result[:warmup_period], "Expected NaN in warmup period for long")
        assert_all_nan(short_result[:warmup_period], "Expected NaN in warmup period for short")


        if len(long_result) > 240:
            assert_no_nan(long_result[240:], "Found unexpected NaN in long values after warmup")
            assert_no_nan(short_result[240:], "Found unexpected NaN in short values after warmup")

    def test_cksp_streaming(self, test_data):
        """Test CKSP streaming functionality - mirrors check_cksp_streaming"""
        params = EXPECTED_OUTPUTS['cksp']['default_params']


        stream = ta_indicators.CkspStream(
            p=params['p'],
            x=params['x'],
            q=params['q']
        )


        stream_long = []
        stream_short = []
        for i in range(len(test_data['close'])):
            result = stream.update(
                test_data['high'][i],
                test_data['low'][i],
                test_data['close'][i]
            )
            if result is None:
                stream_long.append(np.nan)
                stream_short.append(np.nan)
            else:
                long_val, short_val = result
                stream_long.append(long_val)
                stream_short.append(short_val)


        batch_long, batch_short = ta_indicators.cksp(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            p=params['p'],
            x=params['x'],
            q=params['q']
        )


        assert_close(
            stream_long,
            batch_long,
            rtol=1e-8,
            msg="CKSP streaming long values mismatch"
        )
        assert_close(
            stream_short,
            batch_short,
            rtol=1e-8,
            msg="CKSP streaming short values mismatch"
        )

    def test_cksp_batch(self, test_data):
        """Test CKSP batch processing - mirrors batch tests"""

        result = ta_indicators.cksp_batch(
            test_data['high'],
            test_data['low'],
            test_data['close'],
            p_range=(10, 10, 0),
            x_range=(1.0, 1.0, 0.0),
            q_range=(9, 9, 0)
        )


        assert 'long_values' in result
        assert 'short_values' in result
        assert 'p' in result
        assert 'x' in result
        assert 'q' in result


        assert result['long_values'].shape == (1, len(test_data['close']))
        assert result['short_values'].shape == (1, len(test_data['close']))


        assert result['p'][0] == 10
        assert result['x'][0] == 1.0
        assert result['q'][0] == 9


        result = ta_indicators.cksp_batch(
            test_data['high'][:100],
            test_data['low'][:100],
            test_data['close'][:100],
            p_range=(5, 15, 5),
            x_range=(0.5, 1.5, 0.5),
            q_range=(5, 10, 5)
        )


        assert result['long_values'].shape == (18, 100)
        assert result['short_values'].shape == (18, 100)
        assert len(result['p']) == 18
        assert len(result['x']) == 18
        assert len(result['q']) == 18



        assert result['p'][0] == 5
        assert result['x'][0] == 0.5
        assert result['q'][0] == 5


        assert result['p'][-1] == 15
        assert result['x'][-1] == 1.5
        assert result['q'][-1] == 10


        for i, (p, q) in enumerate(zip(result['p'], result['q'])):
            expected_warmup = p + q - 1
            row_data = result['long_values'][i]

            nan_count = 0
            for val in row_data:
                if np.isnan(val):
                    nan_count += 1
                else:
                    break

            assert nan_count >= min(expected_warmup, 100), f"Row {i}: Expected at least {expected_warmup} NaN values, got {nan_count}"

    def test_cksp_batch_metadata(self, test_data):
        """Test CKSP batch metadata and edge cases"""
        close = test_data['close'][:50]
        high = test_data['high'][:50]
        low = test_data['low'][:50]


        result = ta_indicators.cksp_batch(
            high, low, close,
            p_range=(10, 10, 1),
            x_range=(1.0, 1.0, 0.1),
            q_range=(9, 9, 1)
        )

        assert result['long_values'].shape == (1, 50)
        assert len(result['p']) == 1


        result = ta_indicators.cksp_batch(
            high, low, close,
            p_range=(10, 12, 10),
            x_range=(1.0, 1.0, 0),
            q_range=(9, 9, 0)
        )


        assert result['long_values'].shape == (1, 50)
        assert result['p'][0] == 10


if __name__ == '__main__':
    pytest.main([__file__, '-v'])
